### PR TITLE
feat: fixing informer issues

### PR DESCRIPTION
**Pull request by AlexsJones**
This completely reimplements the informer system. 
It now avoids the issues using the cache informer watcher which occasionally ommits errors - to my best reckoning it is due to partial object updates not deserialising properly.

### CODE REVIEW

File: pkg/sync/kubernetes/kubernetes_sync.go [hunk 1]
```diff
@@ -74,34 +77,130 @@ func (k *Sync) Notify(ctx context.Context, c chan<- sync.INotify) {
...
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			if err := deleteFuncHandler(obj, objectKey, c); err != nil {
+				k.Logger.Warn(err.Error())
+			}
+		},
+	})
+
+	ctx, cancel := signal.NotifyContext(context.Background(), os.Interrupt)
```
Comment (james-milligan) [inline pkg/sync/kubernetes/kubernetes_sync.go:131]: The top level `context.Context` from `pkg/runtime/start.go` is already listening to `osInterrupt` signals, this context is passed into the start syncer function and down through notify, i believe lines 131 and 132 can be removed.

File: pkg/sync/kubernetes/kubernetes_sync.go [hunk 2]
```diff
@@ -74,34 +77,130 @@ func (k *Sync) Notify(ctx context.Context, c chan<- sync.INotify) {
...
+}
+
+func createFuncHandler(obj interface{}, object client.ObjectKey, c chan<- sync.INotify) error {
+	var ffObj ffv1alpha1.FeatureFlagConfiguration
+	u := obj.(*unstructured.Unstructured)
+	err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &ffObj)
+	if err != nil {
+		return err
+	}
+	if reflect.TypeOf(ffObj) != reflect.TypeOf(ffv1alpha1.FeatureFlagConfiguration{}) {
```
Comment (james-milligan) [inline pkg/sync/kubernetes/kubernetes_sync.go:144]: I dont believe its possible for this line to fail, is it possible to check for the Kind / Version being populated instead? (an empty map passed as `u` to `runtime.DefaultUnstructuredConverter.FromUnstructured(u, &ffObj)` passes this if statement)

Comment (james-milligan) [reply pkg/sync/kubernetes/kubernetes_sync.go:144]: This is the same for the remaining `FuncHandlers`

File: pkg/sync/kubernetes/ffclientset.go [hunk 3]
```diff
@@ -26,6 +25,14 @@ type FeatureFlagClient struct {
 	ns         string
 }
 
+type FFCInterface interface {
+	FeatureFlagConfigurations(namespace string) Interface
```
Comment (skyerus) [inline pkg/sync/kubernetes/ffclientset.go:29]: `Interface` could use a more informative name. Hasn't been introduced here so happy to leave it but something to consider

File: pkg/sync/kubernetes/kubernetes_sync.go [hunk 4]
```diff
@@ -72,36 +88,121 @@ func (k *Sync) Notify(ctx context.Context, c chan<- sync.INotify) {
...
+}
+
+func deleteFuncHandler(obj interface{}, object client.ObjectKey, c chan<- sync.INotify) error {
+	var ffObj v1alpha1.FeatureFlagConfiguration
+	u := obj.(*unstructured.Unstructured)
+	err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &ffObj)
+	if err != nil {
+		return err
+	}
+	if reflect.TypeOf(ffObj) != reflect.TypeOf(v1alpha1.FeatureFlagConfiguration{}) {
```
Comment (james-milligan) [inline pkg/sync/kubernetes/kubernetes_sync.go:197]: This `if` statement will never fail, it should be updated to match the pattern in `updateFuncHandler` and `createFuncHandler`

File: pkg/sync/kubernetes/kubernetes_sync.go [hunk 5]
```diff
@@ -72,36 +88,121 @@ func (k *Sync) Notify(ctx context.Context, c chan<- sync.INotify) {
...
-		k.Logger.Panic(err.Error())
+		log.Fatalln(err)
 	}
 
-	if err := ffv1alpha1.AddToScheme(scheme.Scheme); err != nil {
-		k.Logger.Panic(err.Error())
-	}
+	resource := v1alpha1.GroupVersion.WithResource("featureflagconfigurations")
+	factory := dynamicinformer.NewFilteredDynamicSharedInformerFactory(clusterClient,
+		time.Minute, corev1.NamespaceAll, nil)
```
Comment (skyerus) [inline pkg/sync/kubernetes/kubernetes_sync.go:112]: What does the `time.Minute` signify here? Could be worth a comment

File: pkg/sync/kubernetes/ffclientset.go (hunk 3 above)
Comment (AlexsJones) [reply pkg/sync/kubernetes/ffclientset.go:29]: Done

File: pkg/sync/kubernetes/kubernetes_sync.go (hunk 5 above)
Comment (AlexsJones) [reply pkg/sync/kubernetes/kubernetes_sync.go]: I added it back as a variable and added some  comments

File: pkg/sync/kubernetes/ffclientset.go [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+	"github.com/open-feature/open-feature-operator/apis/core/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+)
+
+type Interface interface {
```
Comment (skyerus) [inline pkg/sync/kubernetes/ffclientset.go:16]: I meant this interface, is it possible to use a more descriptive name?

Comment (AlexsJones) [reply pkg/sync/kubernetes/ffclientset.go]: ah yes, done

**Review by james-milligan — APPROVED** [review]

**Review by skyerus — APPROVED** [review]

**Pull request merged**